server/src/main/java/org/elasticsearch/tasks/TaskManager.java

```diff
@@ -495,6 +552,56 @@ public boolean hasParent(TaskId parentTaskId) {
         public CancellableTask getTask() {
             return task;
         }
+
+        synchronized void registerChildNode(DiscoveryNode node) {
+            if (banChildren) {
+                throw new TaskCancelledException("The parent task was cancelled, shouldn't start any children tasks");
+            }
+            if (childTasksPerNode == null) {
+                childTasksPerNode = new ObjectIntHashMap<>();
+            }
+            childTasksPerNode.addTo(node, 1);
+        }
+
+        void unregisterChildNode(DiscoveryNode node) {
+            final Runnable runnable;
+            synchronized (this) {
+                if (childTasksPerNode.addTo(node, -1) == 0) {
+                    childTasksPerNode.remove(node);
+                }
+                if (childTasksPerNode.isEmpty()) {
+                    runnable = onEmptyChildNodes;
+                } else {
+                    runnable = null;
+                }
+            }
+            if (runnable != null) {
+                runnable.run();
+            }
+        }
+
+        Set<DiscoveryNode> startBan(Runnable onEmptyChildNodes) {
+            final Set<DiscoveryNode> pendingChildNodes;
+            synchronized (this) {
+                if (banChildren) {
+                    logger.warn("Trying to start ban twice for task " + task.getId());
+                    pendingChildNodes = Collections.emptySet();
```

Comment:
Why use empty set here and not the actual set? Will this not lead to incorrectness?

Reply:
Good catch. Fixed in https://github.com/elastic/elasticsearch/pull/54312/commits/226a5414320e8e4f725a1762e2226af6731a906a